[ETD-48] Add RS485 driver implementation and example app

psurm opened the pull request.

psurm committed:
[ETD-48] Create RS485 driver project as a copy of UART driver project

psurm committed:
[ETD-48] Add Arbiter function to RS485 driver

Arbiter is used to control bytes flow in RS485 architecture. It sends bytes when
it has a token that alows the transfer or buffers bytes otherwise. It also
changes value of PioDriver pin.

psurm committed:
[ETD-48] Add Arbiter implementation

psurm committed:
[ETD-160] Add supporting files for packetizer selection

psurm committed:
[ETD-48] Clean project, remove unused commented out code

psurm committed:
[ETD-48] Fix typos in install script

psurm committed:
[ETD-48] Use Pio calls to control communication direction

psurm committed:
[ETD-48] Update stack size and encoding of IV interfaces

psurm committed:
[ETD-48] Add usage example of RS485 driver

psurm committed:
[ETD-48] Change implementation of Arbiter from C to CPP

psurm committed:
[ETD-48] Add RS485 x86 Linux driver

psurm committed:
[ETD-48] Update XML model

psurm committed:
[ETD-48] Make Deliver/Receive interfaces protected

psurm committed:
[ETD-48] Remove sleep call from Arbiter's interface

Lurkerpas commented on targets/rs485_usage_test/work/f2/C/src/f2.c:
```diff
@@ -0,0 +1,30 @@
+  f2_RI_response(&arg);
+}
+
+void f2_PI_trigger( void )
```
If this interface is unused, it can be safely removed.

psurm replied:
done

Lurkerpas commented on drivers/rs485_linux/work/arbiter/implem/default/CPP/src/arbiter_state.h:
```diff
@@ -0,0 +1,27 @@
+public:
+    inline static const std::chrono::milliseconds listening_window_time = 100ms;
+    inline static const uint8_t open_listening_window_message[] = {
+        0x18, 0xbb, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0x00, 0xaa, 0xc9, 0xf3
```
Why this particular message?

psurm replied:
It is the shortes possible packet that holds 1 byte payload. The payload is 0xFF. As a sender and getter PID the values 0xAA and 0xBB are used taken from the darkest corner of imagination.

Lurkerpas replied:
I don't think that we need a CCSDS wrapping with a payload. We need a payload of one byte within an escaped packet, so basically, we should need 3 bytes -> start, dummy, stop.

Lurkerpas commented on drivers/rs485_linux/work/arbiter/implem/default/CPP/src/arbiter.cc:
```diff
@@ -0,0 +1,107 @@
+      data.message_data = reinterpret_cast<asn1SccPlatformSpecificPointer>(
+      &ctxt_arbiter.queue_buffer[current_head]);
+      data.length = ctxt_arbiter.queue_packets_size[i];
+      arbiter_RI_Deliver(&data);
```
Is this sporadic? Is the TASTE queue size sufficient?

psurm replied:
Interfaces between Arbiter and Driver are protected.

Lurkerpas commented on drivers/rs485_linux/work/arbiter/implem/default/CPP/src/arbiter.cc:
```diff
@@ -0,0 +1,107 @@
+   } else {
+      printf("\tSending window\n");
+      ctxt_arbiter.is_receiving = false;
+      arbiter_send_queued_packets();
```
I would consider sending a single packet from the queue,

psurm replied:
What to do with the rest of packets? Shall we drop them?

Lurkerpas replied:
No, wait for another window. BUT, I will resolve this here and issue another Task to do some improvements.

Lurkerpas commented on drivers/rs485_seds/work/arbiter/implem/default/C/src/arbiter.c:
```diff
@@ -0,0 +1,59 @@
+bool arbiter_is_control_packet_received
+      (const uint8_t *message_data, size_t length)
+{
+   if (length == 13 && message_data[8] == 0xFF) {
```
Why don't we check the entire packet. And if we only need one byte to check,then why do we send 13 bytes?

psurm replied:
We send entire packet because one single byte won't pass escaping and checking CRC. We check only payload, which is just one byte.

psurm replied:
The packet is sent here by Assembler and Assembler uses Escaper, just like UART driver.

psurm committed:
[ETD-48] Remove unused interface from RS485 target project

psurm committed:
[ETD-48] Remove some magic numbers

psurm committed:
[ETD-48] Add contents of linux serial ccsds driver

psurm committed:
[ETD-48] Fix ocarina bus value in rs485_linux driver

psurm committed:
[ETD-48] Add configurable PIO pin do ANS.1

Lurkerpas commented on drivers/rs485_linux/rs485_linuxconfiguration.asn:
```diff
@@ -0,0 +1,18 @@
+    RS485-LINUX-Parity-T    ::= ENUMERATED {even, odd}
+
+    RS485-LINUX-Conf-T ::= SEQUENCE {
+        devname        IA5String (SIZE (1..24)),
```
Trim this to provide only the data we actually use

psurm replied:
done

Lurkerpas commented on drivers/rs485_linux/work/arbiter/implem/default/CPP/src/arbiter_state.h:
```diff
@@ -0,0 +1,37 @@
+#include <atomic>
+#include <chrono>
+
+#define QUEUE_SIZE 1000
```
This might be quite short, and on Linux we have all the memory we need. Let's make it 65536

psurm replied:
done

Lurkerpas commented on drivers/rs485_seds/rs485_sedsconfiguration.asn:
```diff
@@ -0,0 +1,31 @@
+    RS485-SEDS-PioHwas-Pin-T ::= INTEGER(0 .. 31)
+
+    RS485-SEDS-PioHwas-T ::= SEQUENCE {
+        mPortConfig RS485-SEDS-PioHwas-Port-T,
```
Rather port ID and pin ID/number; pinconfig has a different meaning (pullup, peripheral, input, schmidt....(

psurm replied:
done

Lurkerpas commented on drivers/rs485_seds/work/assembler/implem/default/C/src/assembler.c:
```diff
@@ -0,0 +1,166 @@
+
+__attribute__((section(".sdramMemorySection")))
+static Escaper escaper;
+#define ENCODED_PACKET_BUFFER_SIZE 100
```
These buffer sizes are really small. I think we should increase them, at least to 1024, however, 65536 would be better.

psurm replied:
done

Lurkerpas commented on drivers/rs485_seds/work/assembler/implem/default/C/src/assembler.c:
```diff
@@ -0,0 +1,166 @@
+   assembler_RI_UartHwas_ReadByteAsyncCmd_Pi(&readByte);
+}
+
+void assembler_PI_Control
```
Can we rename this to "enable transmission" or "start sending" or "try enable transmission"?

psurm replied:
done

Lurkerpas commented on drivers/rs485_seds/work/assembler/implem/default/C/src/assembler.c:
```diff
@@ -0,0 +1,166 @@
+      assembler_RI_UartHwas_SendByteAsyncCmd_Pi(&sendByteStructure);
+   } else {
+      // No more bytes in queue, send acknowledgment to Arbiter.
+      assembler_RI_ControlAck();
```
Can we rename this to transmission ended or data sent?

psurm replied:
done

psurm committed:
[ETD-48] Fix issues found in review

- increase buffers size;
- rename SEDS arbiter interfaces' names;
- remove legacy, unused ASN config entries;
